File: e2e/elixir/test/concurrent_test.exs
# E2e tests for category: concurrent
defmodule E2e.ConcurrentTest do
  use ExUnit.Case, async: true

  describe "concurrent_basic" do
    test "Concurrent crawling fetches all pages with max_concurrent workers" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(length(result.pages)) == 6
      assert length(result.pages) >= 6
    end
  end

  describe "concurrent_depth_two_fan_out" do
    test "Concurrent depth=2 crawl correctly fans out and deduplicates across levels" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(length(result.pages)) == 4
    end
  end

  describe "concurrent_max_pages_exact" do
    test "Concurrent crawling does not exceed max_pages limit even with high concurrency" do
      result = Kreuzcrawl.scrape!()
      assert length(result.pages) <= 3
    end
  end

  describe "concurrent_partial_errors" do
    test "Concurrent crawl handles partial failures gracefully" do
      result = Kreuzcrawl.scrape!()
      assert length(result.pages) >= 2
    end
  end

  describe "concurrent_respects_max_pages" do
    test "Concurrent crawling respects max_pages limit" do
      result = Kreuzcrawl.scrape!()
      assert length(result.pages) <= 3
    end
  end
end

File: e2e/rust/tests/concurrent_test.rs
//! E2e tests for category: concurrent

use kreuzcrawl::scrape;
use kreuzcrawl::create_engine;

#[tokio::test]
async fn test_concurrent_basic() {
    // Concurrent crawling fetches all pages with max_concurrent workers
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let _ = scrape(&engine, &url).await.expect("should succeed");
    // skipped: field 'pages.length' not available on result type
    // skipped: field 'pages.length' not available on result type
}

#[tokio::test]
async fn test_concurrent_depth_two_fan_out() {
    // Concurrent depth=2 crawl correctly fans out and deduplicates across levels
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let _ = scrape(&engine, &url).await.expect("should succeed");
    // skipped: field 'pages.length' not available on result type
}

#[tokio::test]
async fn test_concurrent_max_pages_exact() {
    // Concurrent crawling does not exceed max_pages limit even with high concurrency
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let _ = scrape(&engine, &url).await.expect("should succeed");
    // skipped: field 'pages.length' not available on result type
}

#[tokio::test]
async fn test_concurrent_partial_errors() {
    // Concurrent crawl handles partial failures gracefully
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let _ = scrape(&engine, &url).await.expect("should succeed");
    // skipped: field 'pages.length' not available on result type
}

#[tokio::test]
async fn test_concurrent_respects_max_pages() {
    // Concurrent crawling respects max_pages limit
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let _ = scrape(&engine, &url).await.expect("should succeed");
    // skipped: field 'pages.length' not available on result type
}
